src/lib/io/include/sol-memmap-storage.h

```diff
@@ -0,0 +1,273 @@
+ * stored map is different from map version. Note that @c _version field
+ * is a @c uint8_t and that versions should start on 1, so Soletta will know
+ * if dealing with a totally new storage.
+ *
+ * @{
+ */
+
+#define MEMMAP_VERSION_ENTRY "_version" /**< Name of property which contains stored map version */
+
+#define SOL_MEMMAP_ENTRY(_name, _offset, _size) \
+    SOL_STR_TABLE_PTR_ITEM(_name, &((struct sol_memmap_entry){.offset = (_offset), .size = (_size) }))
+
+#define SOL_MEMMAP_BOOL_ENTRY(_name, _offset, _bit_offset) \
+    SOL_STR_TABLE_PTR_ITEM(_name, &((struct sol_memmap_entry){.offset = (_offset), .size = 1, .bit_offset = (_bit_offset), .bit_size = 1 }))
+
+#define SOL_MEMMAP_ENTRY_BIT_SIZE(_name, _offset, _size, _bit_offset, _bit_size) \
+    SOL_STR_TABLE_PTR_ITEM(_name, &((struct sol_memmap_entry){.offset = (_offset), .size = (_size), .bit_offset = (_bit_offset), .bit_size = (_bit_size) }))
+
+struct sol_memmap_map {
+    uint8_t version; /**< Version of map. Functions will refuse to read/write on storage if this version and the one storad differs */
+    char *path; /**< Where to find the storage. Under Linux, it is the file mapping the storage, like @c /dev/nvram */
+    struct sol_str_table_ptr entries[]; /**< Entries on map, containing name, offset and size */
+};
+
+struct sol_memmap_entry {
+    size_t offset; /**< Offset of this entry on storage, in bytes. If zero, it will be calculated from previous entry on @c entries array */
+    size_t size; /**< Total size of this entry on storage, in bytes. */
+    uint32_t bit_size; /**< Total size of this entry on storage, in bits. Must be up to <tt>size * 8</tt>. If zero, it will be assumed as <tt>size * 8</tt>. Note that this will be ignored if @c size is greater than 8. */
+    uint8_t bit_offset; /**< Bit offset on first byte. Note that this will be ignored if @c size is greater than 8. */
+};
+
+/**
+ * Writes buffer contents to storage.
+ *
+ * @param name name of property. must be present in one of maps previoulsy
+ * added via @c sol_memmap_add_map (if present in more than one,
+ * behaviour is undefined)
+ * @param buffer buffer that will be written, according to its entry on map.
+ *
+ * return 0 on success, a negative number on failure
+ */
+int sol_memmap_write_raw(const char *name, struct sol_buffer *buffer);
+
+/**
+ * Read storage contents to buffer.
+ *
+ * @param name name of property. must be present in one of maps previoulsy
+ * added via @c sol_memmap_add_map (if present in more than one,
+ * behaviour is undefined)
+ * @param buffer buffer where result will be read into, according to its entry
+ * on map.
+ *
+ * return 0 on success, a negative number on failure
+ */
+int sol_memmap_read_raw(const char *name, struct sol_buffer *buffer);
+
+/**
+ * Add a map to internal list of available maps.
+ *
+ * @param map map to be add.
+ *
+ * @return 0 on success, a negative number on failure.
+ */
+int sol_memmap_add_map(const struct sol_memmap_map *map);
```

Comment:
do we keep a reference to this map? yes. then document that, since it's not obvious :-)

Say the caller needs o keep the memory pointed by `map` alive during the memmap usage.